staging/src/k8s.io/kubectl/pkg/cmd/run/run.go

```diff
@@ -628,9 +629,8 @@ func (o *RunOptions) createGeneratedObject(f cmdutil.Factory, cmd *cobra.Command
 		return nil, err
 	}
 
-	if len(overrides) > 0 {
-		codec := runtime.NewCodec(scheme.DefaultJSONEncoder(), scheme.Codecs.UniversalDecoder(scheme.Scheme.PrioritizedVersionsAllGroups()...))
-		obj, err = cmdutil.Merge(codec, obj, overrides)
+	if mapping.GroupVersionKind.Kind == "Pod" {
```

Comment:
When would this _not_ be a pod? If that can happen, shouldn't we return an error rather than ignoring the override data in that case?

Reply:
`createGeneratedObject` gets called from two places:
1. Inside [`Run()`](https://github.com/kubernetes/kubernetes/blob/d4b6daaf6fc7c1489c5cbbccad3ad2967e4459cf/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L325) which creates the pod
2. Inside [`generateService()`](https://github.com/kubernetes/kubernetes/blob/d4b6daaf6fc7c1489c5cbbccad3ad2967e4459cf/staging/src/k8s.io/kubectl/pkg/cmd/run/run.go#L590) which creates the service when you use the `--expose` arg

Those are the only two cases though.

This conditional is to control whether the override should be applied (Pod) or not (Service).

Previously, I had added a boolean parameter that was passed in to indicate whether to apply the override or not.  @eddiezane suggested in a previous review comment checking Kind == Pod here instead, which I thought was a better approach because that information is available already in this function and it eliminates the need for an extra function parameter.

Let me know if this changes your thoughts above, or if you'd like to see a check in place and return an error, and I can add it to check that it is a Pod or Service and if not return an error.  Also maybe a comment would help explaining why this check is performed.

Reply:
Hmm, I see. What still bothers me is that the fact that this function needs to know the available overrides are not always valid, and the reason for that has to do with its callers and not with something fundamental about what it is using to make the distinction (pod vs service). 

I know it's more verbose, but I'd be inclined to break up this big function into the three things it does, only two of which the Service code path wants, i.e.:

```go
	obj, err := o.buildGeneratedObject(generator, names, params)
	if err != nil {
		return err
	}
	obj, err = o.ApplyOverride(obj, corev1.Pod{})
	if err != nil {
		return err
	}
	runObject, err := o.createGeneratedObject(f, cmd, obj)
	if err != nil {
		return err
	}
```

WDYT of that option?

Even there, it kinda bugs me to hardcode `corev1.Pod{}` at all, since the surrounding code seems to be making an effort to be generic. 

Looking at the `LookupPatchMeta` options available in `strategicpatch`, I'm not sure how to get around that... unless we have openAPI available, maybe? Or there's some way to go from GVK (which we clearly have here) to Go struct? Or we could somehow get this override step right into the part that knows this is a pod, i.e. `func (BasicPod) Generate`?